From 1ac5b2893950ef511ad3d759ee7f595d47fe93ff Mon Sep 17 00:00:00 2001
From: Cody Peter Mello <cody.mello@joyent.com>
Date: Tue, 27 Feb 2018 22:36:50 +0000
Subject: [PATCH] TRITON-75 Allow firewall rules to be applied in order of
 priority TRITON-77 Add support for ESP and AH protocols in firewall rules
 TRITON-117 Update number of ports allowed per firewall rule Reviewed by: Nick
 Zivkovic <nick.zivkovic@joyent.com> Approved by: Nick Zivkovic
 <nick.zivkovic@joyent.com>

---
 .gitignore                       |  2 ++
 deps/jsstyle                     |  2 +-
 docs/index.md                    |  1 +
 lib/rule.js                      |  6 +++++
 lib/update.js                    |  2 +-
 package.json                     |  6 ++---
 test/integration/get.test.js     | 17 ++++++++++++-
 test/integration/global.test.js  |  2 +-
 test/integration/invalid.test.js |  2 +-
 test/integration/list.test.js    | 43 +++++++++++++++++++++++++++++++-
 test/integration/resolve.test.js |  2 +-
 test/integration/update.test.js  |  2 +-
 test/lib/rule.js                 |  2 +-
 test/unit/data/migration.js      |  2 +-
 test/unit/helpers.js             |  4 +--
 test/unit/rule.test.js           |  2 +-
 16 files changed, 81 insertions(+), 16 deletions(-)

diff --git a/.gitignore b/.gitignore
index e14bb32..6798ae8 100644
--- a/.gitignore
+++ b/.gitignore
@@ -2,6 +2,8 @@
 /node_modules
 /tmp
 build
+docs/examples.md
+docs/rules.md
 docs/*.json
 docs/*.html
 docs/examples.md
diff --git a/deps/jsstyle b/deps/jsstyle
index 9600c7e..ccb145b 160000
--- a/deps/jsstyle
+++ b/deps/jsstyle
@@ -1 +1 @@
-Subproject commit 9600c7e56c84f3a74d6e3d70c336e86f7a3e3769
+Subproject commit ccb145bab281b81e1fbebd45b5d2c084c454df5d
diff --git a/docs/index.md b/docs/index.md
index 0097bc2..33acc22 100644
--- a/docs/index.md
+++ b/docs/index.md
@@ -157,6 +157,7 @@ FROM or TO side of the rule.
 | fields     | Array of Strings | List of extra fields to return |
 | ip         | String           | Filter: IP                     |
 | owner_uuid | UUID             | Filter: Owner UUID             |
+| protocol   | String           | Filter: Protocol (e.g., "tcp") |
 | subnet     | String           | Filter: Subnet CIDR            |
 | tag        | String           | Filter: Tag                    |
 | vm         | UUID             | Filter: VM UUID                |
diff --git a/lib/rule.js b/lib/rule.js
index 974d925..5aabc8f 100644
--- a/lib/rule.js
+++ b/lib/rule.js
@@ -113,6 +113,7 @@ function ruleFromMoray(raw) {
             (raw.enabled === 'true'),
         objectclass: raw.objectclass,
         parsed: {
+            priority: raw.priority,
             action: raw.action,
             from: [],
             protocol: {
@@ -124,6 +125,10 @@ function ruleFromMoray(raw) {
         version: raw.version
     };
 
+    if (raw.protocol === 'ah' || raw.protocol === 'esp') {
+        data.parsed.protocol.targets = [ 'all' ];
+    }
+
     if (hasKey(raw, 'ports')) {
         // See "Port storage in Moray" for how ports get represented
         data.parsed.protocol.targets = raw.ports.map(function (port) {
@@ -387,6 +392,7 @@ Rule.prototype.rawMoray = function _ruleRawMoray() {
         action: this.action,
         enabled: this.enabled,
         objectclass: Rule.objectclass,
+        priority: this.priority,
         protocol: this.protocol,
         uuid: this.uuid,
         version: this.version,
diff --git a/lib/update.js b/lib/update.js
index 92ea130..3940b5b 100644
--- a/lib/update.js
+++ b/lib/update.js
@@ -17,7 +17,7 @@
 var assert = require('assert-plus');
 var stream = require('fast-messages');
 var util = require('util');
-var uuid = require('node-uuid');
+var uuid = require('uuid');
 
 
 var hasKey = require('jsprim').hasKey;
diff --git a/package.json b/package.json
index 99b6239..6114a07 100644
--- a/package.json
+++ b/package.json
@@ -9,7 +9,7 @@
     "faucet": "0.0.1",
     "istanbul": "0.4.5",
     "eslint": "2.13.1",
-    "eslint-plugin-joyent": "1.1.0",
+    "eslint-plugin-joyent": "~1.3.0",
     "mockery": "1.4.0"
   },
   "dependencies": {
@@ -20,14 +20,14 @@
     "effluent-logger": "git+https://github.com/joshwilsdon/effluent-logger.git#d662f161a07f94045ad2afb45442931511c40e51",
     "extsprintf": "1.0.2",
     "fast-messages": "1.0.1",
-    "fwrule": "1.4.1",
+    "fwrule": "2.0.0",
     "jsprim": "1.3.1",
     "moray": "2.0.1",
     "moray-filter": "1.0.0",
     "tape": "4.2.2",
-    "node-uuid": "1.2.0",
     "sdc-clients": "9.5.0",
     "ufds": "1.2.0",
+    "uuid": "3.2.1",
     "restify": "4.3.0",
     "trace-event": "1.3.0",
     "vasync": "1.6.4",
diff --git a/test/integration/get.test.js b/test/integration/get.test.js
index eec4886..2a7c2ac 100644
--- a/test/integration/get.test.js
+++ b/test/integration/get.test.js
@@ -16,7 +16,7 @@
 
 var test = require('tape');
 var mod_rule = require('../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 
 
 
@@ -44,6 +44,21 @@ var RULES = [
         enabled: true,
         global: true,
         rule: 'FROM any TO tag "foo" = "baz" ALLOW tcp PORT 5010'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[1],
+        rule: 'FROM tag "a" TO tag "b" ALLOW tcp PORT 80 PRIORITY 50'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[1],
+        rule: 'FROM tag "a" TO tag "b" ALLOW ah'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[1],
+        rule: 'FROM tag "a" TO tag "b" ALLOW esp'
     }
 ];
 
diff --git a/test/integration/global.test.js b/test/integration/global.test.js
index 9607f6e..2843d1e 100644
--- a/test/integration/global.test.js
+++ b/test/integration/global.test.js
@@ -16,7 +16,7 @@
 
 var test = require('tape');
 var mod_rule = require('../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 
 
 
diff --git a/test/integration/invalid.test.js b/test/integration/invalid.test.js
index 16f3444..2484b9f 100644
--- a/test/integration/invalid.test.js
+++ b/test/integration/invalid.test.js
@@ -17,7 +17,7 @@
 var test = require('tape');
 var mod_err = require('../../lib/errors');
 var mod_rule = require('../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 
 
 // --- Globals
diff --git a/test/integration/list.test.js b/test/integration/list.test.js
index fc13b1b..bac06fc 100644
--- a/test/integration/list.test.js
+++ b/test/integration/list.test.js
@@ -18,7 +18,7 @@ var test = require('tape');
 var constants = require('../../lib/util/constants');
 var extend = require('xtend');
 var mod_rule = require('../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 
 
 
@@ -71,6 +71,19 @@ var RULES = [
         owner_uuid: OWNERS[1],
         rule: 'FROM (tag "foo" = "bar" OR tag "foo" = "baz") '
             + 'TO tag "side" = "two" ALLOW tcp (PORT 5003 AND PORT 5004)'
+    },
+
+    // OWNERS[2]
+
+    {
+        enabled: true,
+        owner_uuid: OWNERS[2],
+        rule: 'FROM any TO all vms ALLOW ah'
+    },
+    {
+        enabled: true,
+        owner_uuid: OWNERS[2],
+        rule: 'FROM any TO all vms ALLOW esp'
     }
 ];
 
@@ -186,6 +199,34 @@ test('list: parsed fields', function (t) {
 });
 
 
+test('list: IPsec protocols', function (t) {
+    t.plan(2);
+
+    t.test('list: IPsec protocols - AH', function (t2) {
+        mod_rule.list(t2, {
+            params: {
+                owner_uuid: OWNERS[2],
+                protocol: 'ah'
+            },
+            exp: [
+                RULES[7]
+            ]
+        });
+    });
+
+    t.test('list: IPsec protocols - ESP', function (t2) {
+        mod_rule.list(t2, {
+            params: {
+                owner_uuid: OWNERS[2],
+                protocol: 'esp'
+            },
+            exp: [
+                RULES[8]
+            ]
+        });
+    });
+});
+
 
 // --- Teardown
 
diff --git a/test/integration/resolve.test.js b/test/integration/resolve.test.js
index 4da22f4..e3577fa 100644
--- a/test/integration/resolve.test.js
+++ b/test/integration/resolve.test.js
@@ -18,7 +18,7 @@ var test = require('tape');
 var assert = require('assert-plus');
 var fmt = require('util').format;
 var mod_rule = require('../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 var vasync = require('vasync');
 
 
diff --git a/test/integration/update.test.js b/test/integration/update.test.js
index f382d00..cb226c4 100644
--- a/test/integration/update.test.js
+++ b/test/integration/update.test.js
@@ -16,7 +16,7 @@
 
 var test = require('tape');
 var mod_rule = require('../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 var vasync = require('vasync');
 
 
diff --git a/test/lib/rule.js b/test/lib/rule.js
index 9c01dfe..9bc7c9a 100644
--- a/test/lib/rule.js
+++ b/test/lib/rule.js
@@ -23,7 +23,7 @@ var ifErr = common.ifErr;
 var mod_client = require('./client');
 var mod_err = require('../../lib/errors');
 var mod_trunc = require('./trunc');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 var mod_log = require('./log');
 var mod_vasync = require('vasync');
 
diff --git a/test/unit/data/migration.js b/test/unit/data/migration.js
index a3b6a68..8011149 100644
--- a/test/unit/data/migration.js
+++ b/test/unit/data/migration.js
@@ -16,7 +16,7 @@
 
 var assert = require('assert-plus');
 var fwrule = require('fwrule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 
 var fmt = require('util').format;
 
diff --git a/test/unit/helpers.js b/test/unit/helpers.js
index 9492690..7abdefb 100644
--- a/test/unit/helpers.js
+++ b/test/unit/helpers.js
@@ -23,7 +23,7 @@ var mod_common = require('../lib/common');
 var mod_log = require('../lib/log');
 var mod_moray = require('../../lib/moray');
 var mod_update = require('../../lib/update');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 var moray_sandbox = require('moray-sandbox');
 var os = require('os');
 
@@ -33,7 +33,7 @@ var os = require('os');
 
 
 var CUR_IP = 1;
-var FWRULE_VERSION = process.env.FWRULE_VERSION || 3;
+var FWRULE_VERSION = process.env.FWRULE_VERSION || 4;
 var MOCKS_ENABLED = false;
 var MULTI_SUITE_RUN = false;
 var OWNER_UUID = mod_uuid.v4();
diff --git a/test/unit/rule.test.js b/test/unit/rule.test.js
index ddbad18..f4dc25a 100644
--- a/test/unit/rule.test.js
+++ b/test/unit/rule.test.js
@@ -16,7 +16,7 @@
 
 var test = require('tape');
 var mod_rule = require('../../lib/rule');
-var mod_uuid = require('node-uuid');
+var mod_uuid = require('uuid');
 var util = require('util');
 var util_ip = require('../../lib/util/ip');
 
-- 
2.21.0

